docs/azure/sdk/thread-safety.md

```diff
@@ -0,0 +1,121 @@
+description: Learn about thread safety in Azure SDK client objects and how this design impacts client lifetime management.
+ms.date: 05/24/2021
+ms.author: pakrym
+author: pakrym
+---
+
+# Thread safety and client lifetime management for Azure SDK objects
+
+This article helps you understand thread safety issues when using the Azure SDK. It also discusses how the design of the SDK impacts client lifetime management. You'll learn why it's unnecessary to dispose of Azure SDK client objects.
+
+## Thread safety
+
+All Azure SDK client objects are thread-safe and independent of each other. This design ensures that reusing client instances is always safe, even across threads. For example, the following code launches multiple tasks but is thread safe:
+
+```csharp
+var client = new SecretClient(new Uri("<secrets_endpoint>"), new DefaultAzureCredential());
+
+foreach (var secretName in secretNames)
+{
+    // Using clients from parallel threads
+    Task.Run(() => Console.WriteLine(client.GetSecret(secretName).Value));
+}
+```
+
+Model objects used by SDK clients, whether input or output models, are *not* thread-safe by default. Most use cases involving model objects only involve a single thread, making the cost of implementing synchronization as a default behavior too high for these objects. The following sample illustrates a bug where accessing a model from multiple threads could cause an undefined behavior.
+
+```csharp
+KeyVaultSecret newSecret = client.SetSecret("secret", "value");
+
+foreach (var tag in tags)
+{
+    // Don't use model type from parallel threads
+    Task.Run(() => newSecret.Properties.Tags[tag] = CalculateTagValue(tag));
+}
+
+client.UpdateSecretProperties(newSecret.Properties);
+
+```
+
+To access the model from different threads, you must implement your own synchronization code. For example:
+
+```csharp
+KeyVaultSecret newSecret = client.SetSecret("secret", "value");
+
+foreach (var tag in tags)
+{
+    Task.Run(() =>
+    {
+        lock (newSecret)
+        {
+            newSecret.Properties.Tags[tag] = CalculateTagValue(tag);
+        }
+    );
+}
+
+client.UpdateSecretProperties(newSecret.Properties);
+```
+
+## Client lifetime
+
+Because Azure SDK clients are thread-safe, there's no reason to construct multiple SDK client objects for a given set of constructor parameters. Therefore, you should treat Azure SDK client objects as singletons once constructed. This is most commonly implemented by registering Azure SDK client objects as singletons in the app's Inversion of Control (IoC) container and using dependency injection (DI) to obtain references to the SDK client object. The following example shows how this is done in an ASP.NET Core app with the built-in IoC container.
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddControllersWithViews();
+
+    var blobServiceClient = new BlobServiceClient(new Uri("<secrets_endpoint>"), new DefaultAzureCredential());                   
```

Comment:
Wrap this line to prevent horizontal scrolling in docs.

```suggestion
    var blobServiceClient = new BlobServiceClient(
        new Uri("<secrets_endpoint>"), new DefaultAzureCredential());
```